Avoid DDE in narrow with unbacked start

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #166379
* __->__ #166361

Slice knows how to handle unbacked start, we do not need to offset start before calling slice, we can leave it for slice.
The only edge case is when start<0 and start+length ==0 in that case slice and narrow would deviate, 
for that case we shall pass dim_size instead of start+length 

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

laithsakka committed:
Avoid DDE in narrow with unbacked start

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid DDE in narrow with unbacked start"


Slice knows how to handle unbacked start, we do not need to offset start before calling slice, we can leave it for slice
but we have to write the check carefully.  

[ghstack-poisoned]

laithsakka commented:
Jinx lol

laithsakka committed:
Update on "Avoid DDE in narrow with unbacked start"


Slice knows how to handle unbacked start, we do not need to offset start before calling slice, we can leave it for slice
but we have to write the check carefully.  

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid DDE in narrow with unbacked start"


Slice knows how to handle unbacked start, we do not need to offset start before calling slice, we can leave it for slice
but we have to write the check carefully.  

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid DDE in narrow with unbacked start"


Slice knows how to handle unbacked start, we do not need to offset start before calling slice, we can leave it for slice
but we have to write the check carefully.  

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid DDE in narrow with unbacked start"


Slice knows how to handle unbacked start, we do not need to offset start before calling slice, we can leave it for slice.
The only edge case is when start<0 and start+length ==0 in that case slice and narrow would deviate, 
for that case we shall pass dim_size instead of start+length 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid DDE in narrow with unbacked start"


Slice knows how to handle unbacked start, we do not need to offset start before calling slice, we can leave it for slice.
The only edge case is when start<0 and start+length ==0 in that case slice and narrow would deviate, 
for that case we shall pass dim_size instead of start+length 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid DDE in narrow with unbacked start"


Slice knows how to handle unbacked start, we do not need to offset start before calling slice, we can leave it for slice.
The only edge case is when start<0 and start+length ==0 in that case slice and narrow would deviate, 
for that case we shall pass dim_size instead of start+length 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid DDE in narrow with unbacked start"


Slice knows how to handle unbacked start, we do not need to offset start before calling slice, we can leave it for slice.
The only edge case is when start<0 and start+length ==0 in that case slice and narrow would deviate, 
for that case we shall pass dim_size instead of start+length 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid DDE in narrow with unbacked start"


Slice knows how to handle unbacked start, we do not need to offset start before calling slice, we can leave it for slice.
The only edge case is when start<0 and start+length ==0 in that case slice and narrow would deviate, 
for that case we shall pass dim_size instead of start+length 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid DDE in narrow with unbacked start"


Slice knows how to handle unbacked start, we do not need to offset start before calling slice, we can leave it for slice.
The only edge case is when start<0 and start+length ==0 in that case slice and narrow would deviate, 
for that case we shall pass dim_size instead of start+length 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid DDE in narrow with unbacked start"


Slice knows how to handle unbacked start, we do not need to offset start before calling slice, we can leave it for slice.
The only edge case is when start<0 and start+length ==0 in that case slice and narrow would deviate, 
for that case we shall pass dim_size instead of start+length 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

aorenste commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1710,19 +1711,46 @@ Tensor narrow_symint(
+    // This says at runtime: if start < 0 and end == 0, then pass curr_size
+    // instead of 0.
+
+    auto use_different = start.sym_lt(0).sym_and(end.sym_eq(0)).toSymInt();
```
nit: I don't like this name, but don't have a better suggestion, lol.

laithsakka commented on torch/utils/_sympy/printers.py:
```diff
@@ -327,6 +345,24 @@ def _print_Where(self, expr: sympy.Expr) -> str:
+                if result is None:
+                    result = expr_str
+                else:
+                    result = f"{cond_str} ? {expr_str} : {result}"
```
i think the code coouple lines above already add parenthesize to result we prob ok

laithsakka commented:
@pytorchbot merge

atalman commented:
@pytorchmergebot revert -c nosignal -m "examples/models/llama/tests/test_export_llama_lib.py::ExportLlamaLibTest::test_has_expected_ops_and_op_counts [GH job link](https://github.com/pytorch/pytorch/actions/runs/18993202115/job/54257916041) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/1aef88c72d3aef629b20e97a188c9dc4bab46a1a)"

laithsakka commented:
> @pytorchbot successfully started a revert job. Check the current status [here](https://github.com/pytorch/pytorch/actions/runs/19001947910). Questions? Feedback? Please reach out to the [PyTorch DevX Team](https://github.com/pytorch/pytorch/wiki/Dev-Infra-Office-Hours)
```
test_has_expected_ops_and_op_counts
``` 
Are u sure this fail due to my PR?

laithsakka committed:
Update on "Avoid DDE in narrow with unbacked start"


Slice knows how to handle unbacked start, we do not need to offset start before calling slice, we can leave it for slice.
The only edge case is when start<0 and start+length ==0 in that case slice and narrow would deviate, 
for that case we shall pass dim_size instead of start+length 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 aditew01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

laithsakka commented:
rebase on main i dont see hoe the faluire is related to my change lol

laithsakka commented:
<img width="747" height="244" alt="Screenshot 2025-11-02 at 12 39 35 AM" src="https://github.com/user-attachments/assets/63fb0c81-c397-42b4-9814-415c2b14a7da" />

cc @atalman

laithsakka committed:
Update on "Avoid DDE in narrow with unbacked start"


Slice knows how to handle unbacked start, we do not need to offset start before calling slice, we can leave it for slice.
The only edge case is when start<0 and start+length ==0 in that case slice and narrow would deviate, 
for that case we shall pass dim_size instead of start+length 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 aditew01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge

laithsakka committed:
Update on "Avoid DDE in narrow with unbacked start"


Slice knows how to handle unbacked start, we do not need to offset start before calling slice, we can leave it for slice.
The only edge case is when start<0 and start+length ==0 in that case slice and narrow would deviate, 
for that case we shall pass dim_size instead of start+length 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 aditew01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge

laithsakka committed:
Update on "Avoid DDE in narrow with unbacked start"


Slice knows how to handle unbacked start, we do not need to offset start before calling slice, we can leave it for slice.
The only edge case is when start<0 and start+length ==0 in that case slice and narrow would deviate, 
for that case we shall pass dim_size instead of start+length 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 aditew01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid DDE in narrow with unbacked start"


Slice knows how to handle unbacked start, we do not need to offset start before calling slice, we can leave it for slice.
The only edge case is when start<0 and start+length ==0 in that case slice and narrow would deviate, 
for that case we shall pass dim_size instead of start+length 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 aditew01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid DDE in narrow with unbacked start"


Slice knows how to handle unbacked start, we do not need to offset start before calling slice, we can leave it for slice.
The only edge case is when start<0 and start+length ==0 in that case slice and narrow would deviate, 
for that case we shall pass dim_size instead of start+length 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 aditew01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid DDE in narrow with unbacked start"


Slice knows how to handle unbacked start, we do not need to offset start before calling slice, we can leave it for slice.
The only edge case is when start<0 and start+length ==0 in that case slice and narrow would deviate, 
for that case we shall pass dim_size instead of start+length 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 aditew01 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge
